### PR TITLE
Data science

### PR DESCRIPTION
**Pull request by rcaileff**
Second iteration

### CODE REVIEW

File: data_science/spec/lib/cohort_spec.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
+require 'spec_helper'
+require 'cohort'
+
+describe Cohort do
+  let(:cohort) { Cohort.new(name: "A", views: 20, conversions: 10) }
```
Comment (paulghaddad) [inline data_science/spec/lib/cohort_spec.rb]: In the future, you can use tools like Factory Girl to build your data.

File: data_science/spec/lib/cohort_spec.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+  # happy path tests
+  describe "#non_conversions" do
+    it "is views minus conversions" do
+      expect(uneven_cohort.non_conversions).to eq(12)
+    end
+  end
+
+  describe "#conversion_rate_min" do
+    it "is right" do
```
Comment (paulghaddad) [inline data_science/spec/lib/cohort_spec.rb]: Your `it` statements are pretty vague. Take a look at them and see if you can make them more descriptive, such as:

`it "calculates the minimum conversion rate" do`

File: data_science/spec/lib/cohort_spec.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+      expect(cohort.conversion_rate_min).to be_within(0.01).of(0.28)
+    end
+  end
+
+  describe "#conversion_rate_max" do
+    it "is right value" do
+      expect(cohort.conversion_rate_max).to be_within(0.01).of(0.72)
```
Comment (paulghaddad) [inline data_science/spec/lib/cohort_spec.rb]: Good use of the `be_within` matcher.

File: data_science/spec/lib/cohort_spec.rb [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  describe "#conversion_rate_midpoint" do
+    it "it is right" do
+      expect(cohort.conversion_rate_midpoint).to be_within(0.01).of(0.5)
+    end
+  end
+
+  describe "#to_s" do
+    context "when the cohort has a name" do
+      let(:cohort_a) { Cohort.new(name: "A", views: 20, conversions: 18) }
+      it "correctly converts itself to a string" do
```
Comment (paulghaddad) [inline data_science/spec/lib/cohort_spec.rb]: Need a space above this line.

File: data_science/spec/lib/data_science_spec.rb [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+
+describe DataScience do
+  # happy path tests
+  it "outputs to std_out" do
+    expect { subject.report }.to output.to_stdout
+  end
+
+  context "when B is significantly better than A" do
+    let(:researcher) { RoboResearcherFactory.create([250, 50, 250, 100]) }
+    it "reports B is significantly better" do
```
Comment (paulghaddad) [inline data_science/spec/lib/data_science_spec.rb]: Space above needed.

File: data_science/spec/lib/data_science_spec.rb [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+      let(:researcher) { RoboResearcherFactory.create([20, 10, 20, 9]) }
+      it "reports there is no significant difference" do
+        data_science = DataScience.new(researcher: researcher)
+        expect(data_science.conclude).to include("no significant difference")
+      end
+    end
+  end
+end
+
+module RoboResearcherFactory
```
Comment (paulghaddad) [inline data_science/spec/lib/data_science_spec.rb]: Consider extracting this into a method in a `support.rb` file.

File: data_science/spec/lib/data_set_spec.rb [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    it "correctly populates data" do
+      data_set = DataSet.new(file_name: data_file)
+      expect(data_set.views("B")).to eql(2)
+      expect(data_set.conversions("B")).to eql(2)
+    end
+  end
+
+  describe "#cohorts" do
+    it "lists the cohorts" do
+      expect(data_set.cohorts).to match_array(%w(B C))
```
Comment (paulghaddad) [inline data_science/spec/lib/data_set_spec.rb]: Good matcher here!

File: data_science/spec/lib/data_set_spec.rb [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+        expect(data_set.conversions("C")).to eq(0)
+      end
+    end
+  end
+
+  # sad path tests
+  context "when reading an empty file" do
+    let(:data_file) { "spec/spec_data/empty.json" }
+    it "has no data" do
+      data_set = DataSet.new(file_name: data_file)
```
Comment (paulghaddad) [inline data_science/spec/lib/data_set_spec.rb]: This is an overall comment. I think it is best to think in terms of phases of a test. This test has a setup phase and a verification phase. I think having a line break here makes it more readable.

See this article for reference:

https://robots.thoughtbot.com/four-phase-test

File: data_science/spec/spec_data/list_of_two_hashes.json [hunk 9]
```diff
@@ -0,0 +1 @@
+[{"date":"2014-03-20","cohort":"B","result":1}, {"date":"2014-03-21","cohort":"B","result":1}]
```
Comment (paulghaddad) [inline data_science/spec/spec_data/list_of_two_hashes.json]: Good job testing against a small data set.

File: data_science/lib/cohort.rb [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def bad_numeric_arg_raise(arg_name)
+    raise(ArgumentError, "#{arg_name} must be a non-negative integer")
+  end
+
+  def bad_numeric_arg?(arg_value)
+    return true unless arg_value.respond_to?(:round)
+    return true if arg_value < 0
+    return true unless arg_value.round == arg_value
```
Comment (zsyed1) [inline data_science/lib/cohort.rb]: Interesting.

File: data_science/lib/cohort.rb [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+  end
+
+  def conversion_rate_interval
+    return [0, 0] if views == 0
```
Comment (zsyed1) [inline data_science/lib/cohort.rb]: Good!

File: data_science/lib/data_science.rb [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+    standardized_significance = stats_standard_significance(significance)
+    "Cohort #{best.name} is better with 95% confidence.  The difference is " \
+    "significant with #{standardized_significance}% confidence."
+  end
+
+  def stats_standard_significance(significance)
+    return 99.99 if significance >= 99.99
+    return 99.9 if significance >= 99.9
+    return 99 if significance >= 99
+    significance.round
```
Comment (zsyed1) [inline data_science/lib/data_science.rb]: :smile:

File: data_science/lib/data_set.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+class BadDataFile < StandardError
+end
+
+class DataSet
+  attr_reader :file_name, :data
+
+  def initialize(data_points: [], file_name: nil)
+    @file_name = file_name
+    @data = {}
+    data_points += read_json if file_name
```
Comment (zsyed1) [inline data_science/lib/data_set.rb]: Why store all the data-points. You only need to track number of conversions and non-conversions (or equivalently the total views).

File: data_science/lib/robo_researcher.rb [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+  def abtest_format(cohort)
+    { yes: cohort.conversions, no: cohort.non_conversions }
+  end
+
+  def p_to_percentage(p)
+    (1 - p) * 100
+  end
+
+  def best_cohort
+    rate0 = cohorts[0].conversion_rate_midpoint
```
Comment (zsyed1) [inline data_science/lib/robo_researcher.rb]: How do you know `cohorts[0]` and `cohorts[1]` exist?

File: data_science/spec/lib/cohort_spec.rb [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+        expect(description).to include("0 samples")
+        expect(description).to include("0 conversions")
+        expect(description).to include("conversion rate")
+        expect(description).to include("95% confidence")
+      end
+    end
+  end
+  # bad path tests
+  it "raises an error if views is nil" do
+    expect { Cohort.new(name: "F", views: nil, conversions: 0) }.to(
```
Comment (zsyed1) [inline data_science/spec/lib/cohort_spec.rb]: I like how you keep the `Cohort` creating and the error in the same block. Others may prefer these go into a `let`. I find your way easier to read.

File: data_science/spec/spec_helper.rb [hunk 16]
```diff
@@ -0,0 +1,9 @@
+RSpec.configure do |config|
+  config.expect_with :rspec do |expectations|
+    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
+  end
+
+  config.mock_with :rspec do |mocks|
+    mocks.verify_partial_doubles = true
+  end
+end
```
Comment (zsyed1) [inline data_science/spec/spec_helper.rb]: You write very concise code that reads very well.
A lot of your tests are around data-integrity - incredibly thorough.
Very well done!

Zamir

**Comment by paulghaddad** [comment]
@rcaileff Good job Rachel. You are good to move on.

**Pull request closed without merging**